pycloudlib/ibm/instance.py

```diff
@@ -477,37 +477,206 @@ def __init__(
+
+        Returns:
+            A floating IP dict
+        """
+        name = name or f"{self.name}-fi"
+        proto = {
+            "name": name,
+            "resource_group": {"id": self._instance["resource_group"]["id"]},
+            "zone": {"name": self.zone},
+        }
+        floating_ip = self._client.create_floating_ip(proto).get_result()
+        self._created_floating_ips.append(floating_ip)
+        self._log.info("Floating ip created: %s", {floating_ip["name"]})
+        return floating_ip
+
+    def _choose_from_existing_floating_ips(
+        self,
+        name_includes="default-floating-ip",
+    ) -> dict:
+        """
+        Choose a random floating IP from existing floating IPs via substring match.
+
+        Args:
+            name_includes: string, name to filter floating IPs by
+
+        Returns:
+            A floating IP object
+
+        Raises:
+            IBMException: If no floating IPs are found matching the substring
+            IBMCapacityException: If all floating IPs matching the substring are already in use
+        """
+        floating_ips = list(
+            _iter_resources(
+                self._client.list_floating_ips,
+                resource_name="floating_ips",
+                # Select that match the desired name
+                filter_fn=lambda ip: name_includes in ip["name"]
+                and ip.get("zone")
+                and ip["zone"]["name"] == self.zone,
+            )
+        )
+        if not floating_ips:
+            raise IBMException(
+                f"No floating IPs found matching substring {name_includes}"
+            )
+        # filter out floating ips that are already associated with an instance
+        floating_ips = [ip for ip in floating_ips if "target" not in ip]
+        if not floating_ips:
+            raise IBMCapacityException(
+                f"All floating IPs matching substring {name_includes}",
+                "are already in use.",
+            )
+
+        # pick random floating ip to help reduce contention
+        floating_ip = random.choice(floating_ips)
+
+        self._log.info(
+            "Found existing floating ip '%s' with address %s",
+            floating_ip["name"],
+            floating_ip["address"],
+        )
+        return floating_ip
+
+    def attach_floating_ip(
+        self,
+        floating_ip_substring: Optional[str],
+    ) -> Optional[dict]:
+        """
+        Attach a floating IP to this instance.
+
+        If `floating_ip_substring` is given, it will try to attach an existing
+        floating IP that contains the substring in its name. If not, it will
+        create a new floating IP and attach it.
+
+        Args:
+            floating_ip_substring: optional substring to find existing floating IPs by.
+                If not given, ta new floating IP will be created and attached.
+
+        Returns:
+            A floating IP dict if successful, None otherwise
```

Comment:
Do we really want to return the floating IP dict? That dict is a dict that directly comes from the ibm sdk library and:

1. It is unspecified from a pycloudlib perspective.
2. If we return it here, then we are coupling pycloudlib's api with the ibm-sdk one. And if they break, pycloudlib breaks.

If this dict is needed to be exposed for any reason, then I would probably suggest creating a proper class to present it in our API and translate the raw dict response to the class. WDYT?

Reply:
This is honestly a good a point. Thanks for bringing that up. I'll just change it to return the ID or the name or something.